extensions/2.0/Khronos/KHR_materials_transmission/README.md

```diff
@@ -0,0 +1,129 @@
+</figure>
+
+## Blending
+The glTF `blendMode` is used for alpha-as-coverage, NOT for physically-based transparency (i.e. this extension). If alpha-as-coverage is not being used, the blend mode of the material should be set to "OPAQUE" even though it is transparent. Again, it's helpful to think of alpha-as-coverage as whether the physical surface is there or not. `transmission` applies to the surface material that exists.
+Note that alpha-as-coverage can still be used along with transmission as shown in the example below.
+<figure>
+  <img src="./figures/TransmissionWithMask.png"/>
+<figcaption><em>Alpha coverage and optical transparency can be used at the same time so that some areas of a surface are transparent while others disappear entirely.</em></figcaption>
+</figure>
+
+## Transparent Metals
+Metals effectively absorb all refracted light (light that isn't reflected), preventing transmission.
+The metallic parameter of a glTF material effectively scales the `baseColor` of the material toward black while, at the same time scaling the F0 (reflectivity) value towards 1.0. This makes the material opaque for metallic values of 1.0 because transmitted light is attenuated out by absorption. Therefore, for a material with `metallicFactor=1.0`, the value of `transmissionFactor` doesn't matter.
+
+## Implementing Transmission ##
+From the core [glTF BRDF](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#appendix-b-brdf-implementation), we have:
+
+*f* = *f*<sub>*diffuse*</sub> + *f*<sub>*specular*</sub>
+
+*f*<sub>*diffuse*</sub> = (1 - *F*) * *diffuse*
+
+*f*<sub>*specular*</sub> = *F* * *G* * *D* / (4 * dot(*N*, *L*) * dot(*N*, *V*)), where *F* is the Surface Reflection Ratio.
+
+We will now add an additional term for the transmitted light:
+
+*f* = *f*<sub>*diffuse*</sub> + *f*<sub>*specular*</sub> + *f*<sub>*transmission*</sub>
+
+*f*<sub>*transmission*</sub> = (1 - *F*) * *T* * *D<sub>T</sub>* * *baseColor*
+
+where *T* is the transmission percentage defined by this extension's `transmission` and `transmissionTexture` properties and *D<sub>T</sub>* is the distribution function for the transmitted light. The distribution function is the same Trowbridge-Reitz model used by specular reflection except sampled along the view vector rather than the reflection. The *baseColor* factor causes the transmitted light to be tinted by the surface.
+
+Light that penetrates a surface and is transmitted will not be diffusely reflected so we also need to modify the diffuse calculation to account for this.
+*f*<sub>*diffuse*</sub> = (1 - *F*) * (1 - *T*) * *diffuse*
+
+Optical transparency does not require any changes whatsoever to the specular term. So this can all be rearranged so that the transmitted light amounts to a modifaction of the diffuse lobe.
+*f* = (1 - *F*) * (*T* * *D<sub>T</sub>* + (1 - *T*) * *diffuse*) + *f*<sub>*specular*</sub>
+
+## Implementation Notes
```

Comment:
I'm leaning towards that we should require a painter algorithm (depth sorting) of triangles - in my opinion this should be normative.  
Making depth sorting of triangles normative will make a huge step towards consistent result across viewers (bar blurring and possibly other effects)
Without it whole objects may pop in and out of existence......

However, I'm not sure how to approach blurring since I think this will be the most 'heavy' effect to implement for multiple layers/objects.

Reply:
Depth sorting of triangles is not going to be reasonable on all platforms due to the number of passes required (or having the ability to use linked-lists of fragments which isn't possible in WebGL). Besides, I think rendering techniques are out of scope of glTF. Yes, there will be differences between renderers but as long as they do the blending in the way described, I think that's enough.
Blurring of the transmission is pretty cheap if they just sample the mips of the background texture. The background render texture will probably be required anyway to do the appropriate blending.

Reply:
I agree that sorting triangles is prohibitively expensive. Maybe a dedicated model viewer (note: model viewers are a subset of glTF clients) can do that when the camera and object are stationary, but it's not a requirement we should have in the extension spec.

But on a similar point — the core specification says this about alpha blending:

> Whether depth value is written or whether to sort is up to the implementation.

In retrospect, I think we could have made a normative requirement that depth _not_ be written for alpha blend materials. Does that same decision apply to transmission? Or is this not an issue, because transmissive objects are treated as opaque (or rendered in an entirely separate pass)?

Reply:
> Depth sorting of triangles is not going to be reasonable on all platforms due to the number of passes required (or having the ability to use linked-lists of fragments which isn't possible in WebGL).

I agree that if all layers have to support blurring an engine could require multiple passes (render to texture, minify then sample) which could become a performance bottleneck.
However - 'just' sorting the triangles of transmission objects could be done in a single pass if effects needing to sample the objects behind are left out.
This would leave us with viewers that would at least display transparent objects in the correct order and visibility.

Yes, the blurring itself is quite cheap - it's having to render out to texture, forcing multiple passes, that will be mostly heavy.

I also do not like us to be so tied to WebGL - just because something can not be done in WebGL should, in my opinion, not mean that we leave it out.

Reply:
Hope "blurring" is just in short we do it physically - as far as possible - correct.

Reply:
> Hope "blurring" is just in short we do it physically - as far as possible - correct.

Sure, that's what I mean anyway. :-)

The point I'm trying to make is that we should require a minimum of visibility through triangles using transmission.
Meaning that not only the background should show through - any other primitives (that may also use transmission) should also be visible.
However, what effects to apply to each of these primitives is optional - for instance the roughness that will result in blurring (since will require multiple passes on todays hardware)

Reply:
> I also do not like us to be so tied to WebGL - just because something can not be done in WebGL should, in my opinion, not mean that we leave it out.

I agree with this in principle but WebGL will have a huge influence since we want glTF to be viewable on the web.

Reply:
I don't think this is a WebGL problem. Requiring that implementations sort triangles, in realtime, is extremely burdensome on any platform. Yes, there are cases where it may be possible (i.e. few enough triangles), but we really cannot expect this at the data format level.

Is there a reason these requirements cannot be in the 3D Commerce Certification process, instead?